lib/discover_endpoint.js

```diff
@@ -0,0 +1,304 @@
+    identifiers.accessKeyId = service.config.credentials.accessKeyId
+  }
+  return identifiers;
+}
+
+/**
+ * @api private
+ */
+function marshallCustomeIdentifiersHelper(result, params, shape) {
+  if (!shape || params === undefined || params === null) return;
+  if (shape.type === 'structure' && shape.required && shape.required.length > 0) {
+    util.arrayEach(shape.required, function(name) {
+      var memberShape = shape.members[name];
+      if (memberShape.endpointDiscoveryId === true) {
+        var locationName = memberShape.isLocationName ? memberShape.name : name;
+        result[locationName] = String(params[name]);
+      } else {
+        marshallCustomeIdentifiersHelper(result, params[name], memberShape);
+      }
+    });
+  }
+}
+
+/**
+ * Get customized cache key according to the 'endpointDiscoveryId' trait.
+ * @param [object] request object
+ * @param [object] input shape of the given operation's api
+ */
+function marshallCustomeIdentifiers(request, shape) {
+  var identifiers = {};
+  marshallCustomeIdentifiersHelper(identifiers, request.params, shape);
+  return identifiers;
+}
+
+/**
+ * Call endpoint discovery operation when it's optional.
+ * When endpoint is available in cache then use the cached endpoints. If endpoints
+ * are unavailable then use regional endpoints and call endpoint discovery operation
+ * asynchronously. This is turned off by default.
+ * @param [object] request object
+ * @api private
+ */
+function optionalDisverEndpoint(request) {
+  var service = request.service;
+  var api = service.api;
+  var operationModel = api.operations ? api.operations[request.operation] : undefined;
+  var inputShape = operationModel ? operationModel.input : undefined;
+
+  var identifiers = marshallCustomeIdentifiers(request, inputShape);
+  var cacheKey = getCacheKey(request);
+  cacheKey = util.update(cacheKey, identifiers);
+  var endpoints = AWS.endpointCache.get(cacheKey);
+  if (endpoints && endpoints.length > 0) {
+    //found endpoint record from cache
+    request.httpRequest.updateEndpoint(endpoints[0].Address);
+  } else if (endpoints && endpoints.length === 0) {
+    //endpoint operation is being made but response not yet received
+    return;
+  } else {
+    //endpoint record not in cache. make discovery operation
+    var endpointRequest = service.makeRequest(api.endpointOperation, {
+      Operation: operationModel.name,
+      Identifiers: identifiers,
+    });
+    addApiVersionHeader(endpointRequest);
+    endpointRequest.removeListener('validate', AWS.EventListeners.Core.VALIDATE_PARAMETERS);
+    endpointRequest.removeListener('retry', AWS.EventListeners.Core.RETRY_CHECK);
+    var toHead = true;
+    //if endpoint operation failed, keep retrying every minute.
+    endpointRequest.on('retry', function(resp) {
+      var err = resp.error;
+      err.retryable = true;
+      err.retryDelay = 60000;
+      resp.maxRetries = ENDPOINT_OPERATION_MAX_RETRIES;
```

Comment:
I don't think you want to set your maxRetries here. This is currently set at 60, with a 1 minute retryDelay, meaning if this operation never succeeds it will take an hour before the callback passed to `send` is invoked.

That also means this could leave an application running for up to an hour after their process would have normally ended, which isn't what we want.

I'm also unclear on if we need to wait for the endpoint request to complete before making a request to whichever operation was called. I get that the endpoint request can fail, and if it's optional we should continue, but I don't know if we should force trying to get an endpoint on our first request.